akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSpec.scala

```diff
@@ -0,0 +1,219 @@
+
+        val sub = c.expectSubscription()
+        sub.request(5)
+
+        c.expectNext(1)
+        c.expectNext(2)
+        c.expectNext(3)
+
+        c.expectComplete()
+
+        materialized.future.futureValue(Timeout(3.seconds)) should ===("foo")
+      }
+    }
+
+    "use a materializer without auto-fusing" when {
+      implicit val noFusing = ActorMaterializer(
+        ActorMaterializerSettings(system).withAutoFusing(false))
+      implicit def ec = noFusing.executionContext
+
+      val tooDeepForStack = 50000
+
+      "flattening from a future graph" in assertAllStagesStopped {
+        val g = Source.fromFutureSource(Future {
+          Thread.sleep(2000)
+          Fusing.aggressive((1 to tooDeepForStack).
+            foldLeft(Source.single(42).mapMaterializedValue(_ ⇒ 1))(
+              (f, i) ⇒ f.map(identity)))
+        })
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.futureValue(veryPatient) should ===(1)
+        fut.futureValue(veryPatient) should ===(List(42))
+      }
+
+      "flattening from a completion stage" in assertAllStagesStopped {
+        val future: Future[Graph[SourceShape[Int], Int]] = Future {
+          Fusing.aggressive((1 to tooDeepForStack).
+            foldLeft(Source.single(43).mapMaterializedValue(_ ⇒ 1))(
+              (f, i) ⇒ f.map(identity)))
+        }
+        val stage: CompletionStage[Graph[SourceShape[Int], Int]] = future.toJava
+        val g = Source.fromSourceCompletionStage(stage)
+
+        val (mat, fut) = g.toMat(Sink.seq)(Keep.both).run()
+        mat.toScala.futureValue(veryPatient) should ===(1)
+        fut.futureValue(veryPatient) should ===(List(43))
+      }
+    }
+
+    "be cancelled before the underlying Future completes" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val promise = Promise[Source[Int, Int]]()
+        val aside = Promise[Int]()
+        val result = Promise[akka.Done]()
+        def futureSource = Source.fromFutureSource(
+          promise.future).map { i ⇒
+          aside.success(i); i // should never occur
+        }.watchTermination[Unit]() {
+          case (_, res) ⇒ result.completeWith(res); ()
+        }
+
+        futureSource.runWith(Sink.cancelled) should ===(NotUsed)
+        result.future.futureValue should ===(akka.Done)
+        aside.future.isCompleted should ===(false)
+      }
+    }
+
+    "fails as the underlying Future is failed" in {
+      implicit def m = materializer
+
+      assertAllStagesStopped {
+        val promise = Promise[Source[Int, Int]]()
+        val result = Promise[akka.Done]()
+        def futureSource = Source.fromFutureSource(promise.future)
+        def sink = Sink.fold[Int, Int](1)(_ * _)
+
+        promise.failure(new Exception("Foo"))
+
+        futureSource.runWith(sink).failed.
+          map(_.getMessage)(m.executionContext).futureValue should ===("Foo")
+      }
+    }
+
+    "applies back-pressure according future completion" in {
```

Comment:
I don't see the issue. Shouldn't the first `expectNext` (before the expectation) ensure that the first value as been consumed and `map`ed (so that the tested `Promise` completed in `map` must be completed at expected point)?